fabric-client-tags-v1/src/client/java/net/fabricmc/fabric/impl/tag/client/DataLoader.java

```diff
@@ -0,0 +1,125 @@
+import org.slf4j.LoggerFactory;
+
+import net.minecraft.tag.TagEntry;
+import net.minecraft.tag.TagFile;
+import net.minecraft.tag.TagKey;
+import net.minecraft.tag.TagManagerLoader;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.RegistryKey;
+
+import net.fabricmc.fabric.api.tag.client.v1.ClientTags;
+import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.api.ModContainer;
+
+public class DataLoader {
+	private static final Logger LOGGER = LoggerFactory.getLogger("fabric-client-tags-v1");
+	/**
+	 * Load a given tag from the available mods into a set of {@code Identifier}s.
+	 * Parsing based on {@link net.minecraft.tag.TagGroupLoader#loadTags(net.minecraft.resource.ResourceManager)}
+	 */
+	public HashSet<Identifier> loadTag(TagKey<?> tagKey) {
+		var tags = new HashSet<TagEntry>();
+		HashSet<Path> tagFiles = getTagFiles(tagKey.registry(), tagKey.id());
+
+		for (Path tagPath : tagFiles) {
+			try (BufferedReader tagReader = Files.newBufferedReader(tagPath)) {
+				JsonElement jsonElement = JsonParser.parseReader(tagReader);
+				TagFile maybeTagFile = TagFile.CODEC.parse(new Dynamic<>(JsonOps.INSTANCE, jsonElement))
+						.result().orElse(null);
+
+				if (maybeTagFile != null) {
+					if (maybeTagFile.replace()) {
+						tags.clear();
+					}
+
+					tags.addAll(maybeTagFile.entries());
+				}
+			} catch (IOException e) {
+				LOGGER.error("Error loading tag: " + tagKey, e);
+			}
+		}
+
+		HashSet<Identifier> ids = new HashSet<>();
+
+		for (TagEntry tagEntry : tags) {
+			tagEntry.resolve(new TagEntry.ValueGetter<>() {
+				@Nullable
+				@Override
+				public Identifier direct(Identifier id) {
+					return id;
+				}
+
+				@Nullable
+				@Override
+				public Collection<Identifier> tag(Identifier id) {
+					TagKey<?> tag = TagKey.of(tagKey.registry(), id);
+					return ClientTags.getOrCreateLocalTag(tag);
+				}
+			}, ids::add);
+		}
+
+		return ids;
+	}
+
+	/**
+	 * @param registryKey the RegistryKey of the TagKey
+	 * @param identifier  the Identifier of the tag
+	 * @return the paths to all tag json files within the available mods
+	 */
+	private HashSet<Path> getTagFiles(RegistryKey<? extends Registry<?>> registryKey, Identifier identifier) {
+		return getTagFiles(TagManagerLoader.getPath(registryKey), identifier);
+	}
+
+	/**
+	 * @return the paths to all tag json files within the available mods
+	 */
+	private HashSet<Path> getTagFiles(String tagType, Identifier identifier) {
+		String tagFile = "data/%s/%s/%s.json".formatted(identifier.getNamespace(), tagType, identifier.getPath());
+		return getResourcePaths(tagFile);
+	}
+
+	/**
+	 * @return all paths from the available mods that match the given internal path
+	 */
+	private HashSet<Path> getResourcePaths(String path) {
+		HashSet<Path> out = new HashSet<>();
+
+		for (ModContainer mod : FabricLoader.getInstance().getAllMods()) {
+			mod.findPath(path).ifPresent(out::add);
```

Comment:
There might be multiple root paths, how does this work with them?

Reply:
Nevermind, we're only querying the tags file directly and it's not supposed to be duplicated across multiple root paths.